### PR TITLE
Add EVM executive logic

### PR DESCRIPTION
**Pull request by JimmyShi22**
1. Add more StateFace api to support all call by executive context
2. Add more MptState api based on new StateFace
3. Add ExtVM logic to support the callback of EVM
4. Add Executive module to support ExtVM calls

### CODE REVIEW

File: libexecutivecontext/StateFace.h [hunk 1]
```diff
@@ -42,51 +86,74 @@ class StateFace
...
-    /// nonempty). These two notions are equivalent after EIP158.
-    virtual bool accountNonemptyAndExisting(Address const& _address) const = 0;
+    /// The hash of the root of our state tree.
+    virtual h256 rootHash() const = 0;
 
-    /// Check if the address is in use.
-    virtual bool addressInUse(Address const& _address) const = 0;
+    /// Commit all changes waiting in the address cache to the DB.
+    /// @param _commitBehaviour whether or not to remove empty accounts during commit.
+    virtual void commit(StateFace::CommitBehaviour _commitBehaviour) = 0;
```
Comment (wheatli) [inline libexecutivecontext/StateFace.h]: Let remove op be a default logic

File: libexecutivecontext/MptState.cpp [hunk 2]
```diff
@@ -32,9 +32,56 @@ namespace eth
 OverlayDB MptState::openDB(
     boost::filesystem::path const& _path, h256 const& _genesisHash, WithExisting _we)
 {
+    std::mutex open_db_lock;
```
Comment (wheatli) [inline libexecutivecontext/MptState.cpp]: no need lock

File: libexecutivecontext/Executive.h [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+#include <libdevcore/easylog.h>
+#include <libethcore/BlockHeader.h>
+#include <libethcore/Common.h>
+#include <libethcore/Transaction.h>
+#include <libevm/VMFace.h>
+
+#include <json/json.h>
+#include <functional>
+
+#include "ExecutionResult.h"
```
Comment (wheatli) [inline libexecutivecontext/Executive.h]: place it to the begining of include list

File: libexecutivecontext/Executive.cpp [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+                         << *boost::get_error_info<errinfo_evmcStatusCode>(_e) << ")\n"
+                         << diagnostic_information(_e);
+            revert();
+            throw;
+        }
+        catch (Exception const& _e)
+        {
+            // TODO: AUDIT: check that this can never reasonably happen. Consider what to do if it
+            // does.
+            LOG(WARNING)
```
Comment (wheatli) [inline libexecutivecontext/Executive.cpp]: to be ERROR

File: libexecutivecontext/Executive.cpp [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+                << diagnostic_information(_e);
+            exit(1);
+            // Another solution would be to reject this transaction, but that also
+            // has drawbacks. Essentially, the amount of ram has to be increased here.
+        }
+        catch (std::exception const& _e)
+        {
+            // TODO: AUDIT: check that this can never reasonably happen. Consider what to do if it
+            // does.
+            LOG(WARNING) << "Unexpected std::exception in VM. Not enough RAM? " << _e.what();
```
Comment (wheatli) [inline libexecutivecontext/Executive.cpp]: to be ERROR

File: libexecutivecontext/ExecutionResult.h [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    OutOfGas,    ///< Ran out of gas executing code of the transaction.
+    OutOfStack,  ///< Ran out of stack executing code of the transaction.
+    StackUnderflow,
+    NonceCheckFail,  // noncecheck ok() ==  false
+    BlockLimitCheckFail,
+    FilterCheckFail,
+    NoDeployPermission,
+    NoCallPermission,
+    NoTxPermission,
+    EthCallIdNotFound,
```
[the remark this answers is not in the document]
Comment (JimmyShi22) [reply libexecutivecontext/ExecutionResult.h]: done

**Pull request merged**